source/common/common/stack_array.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#if !defined(WIN32)
+#include <alloca.h>
+
+#else
+#include <malloc.h>
+#endif
+
+#include <stddef.h>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+
+// This macro is intended to be used as a replacement for variable-length arrays.
+// Note that the StackArray wrapper object will be destructed when it leaves
+// scope, however the memory containing the array won't be deallocated until
+// the function containing the macro returns
+#define STACK_ARRAY(var, type, num) StackArray<type> var(::alloca(sizeof(type) * num), num)
+
+template <typename T> class StackArray {
+public:
+  StackArray(void* buf, size_t num_items) : num_items_(num_items) {
+    RELEASE_ASSERT(buf != nullptr, "StackArray received null pointer");
+    buf_ = static_cast<T*>(buf);
+    for (size_t i = 0; i < num_items_; i++) {
```

Comment:
Sorry that was confusing. This is really two independent suggestions:

1. Use a range-loop for iterating in the ctor and dtor
2. Now there are no references to num_items_ anywhere in your impl except for the implementation of end(), so just store end_ rather than num_items_.